src/solvers/smt2_incremental/smt_sorts.h

```diff
@@ -0,0 +1,91 @@
+// Author: Diffblue Ltd.
+
+/// \file
+/// Data structure for smt sorts.
+/// \note All classes derived from `smt_sortt`, must be listed in smt_sorts.def.
+
+#ifndef CPROVER_SOLVERS_SMT2_INCREMENTAL_SMT_SORTS_H
+#define CPROVER_SOLVERS_SMT2_INCREMENTAL_SMT_SORTS_H
+
+#include <util/irep.h>
+
+#include <type_traits>
+
+class smt_sort_const_downcast_visitort;
+
+class smt_sortt : protected irept
+{
+public:
+  // smt_sortt does not support the notion of an empty / null state. Use
+  // optionalt<smt_sortt> instead if an empty sort is required.
+  smt_sortt() = delete;
+
+  using irept::pretty;
+
+  bool operator==(const smt_sortt &) const;
+  bool operator!=(const smt_sortt &) const;
+
+  void accept(smt_sort_const_downcast_visitort &) const;
+  void accept(smt_sort_const_downcast_visitort &&) const;
+
+  /// \brief Class for adding the ability to up and down cast smt_sortt to and
+  ///   from irept. These casts are required by other irept derived classes in
+  ///   order to store instances of smt_sortt inside them.
```

Comment:
Why not just publicly inherit from `irept`?

Reply:
Because I want to restrict the ability to cast to and from `irept`, to the inside of the storage classes. I see the fact that each of these inherit from `irept` to be an implementation detail. Ideally, there should be no need to upcast to `irept` as part of the algorithms which live outside the storage classes. Keeping variables declared as `smt_termt` or `smt_sortt` rather than `irept` where possible give us more specific compile time type information.

Reply:
Wouldn't your rationale then suggest using an `irept` member instead of inheriting?

Reply:
> Wouldn't your rationale then suggest using an `irept` member instead of inheriting?

Yes, I could have done that. I could also have implemented this without using `irept` at all and handled lifetime using `std::unique_ptr` or some such. By using `irept` with inheritance this code retains a greater semblance to other parts of the code base. This should aid cbmc developers familiar with other parts of the code base to transition to maintaining or improving this part and it should aid with reusing the existing `pretty`, `operator==` and hashing algorithms defined for `irept`. Otherwise we would need to write more new code to define these pieces of functionality for the new data structures.